### ISSUE TITLE
Filament retraction out of nozzle at start

### ISSUE THREAD

**Issue by delwinbest** [photodude/BIBO2-PrusaSlicer#1]
Hi Walt,
One of the GCode start commands on the BIBO2 E1 right profile is causing a small problem. At times the filament will retract out of the nozzle at the start of a print. 

I can recreate this by:
1. Using Octoprint and the LCD, Before print, extrude between 20 and 50 mm of filament (I do this to ensure there is no previously cooled brittle filament in the head).
2. Load up gcode and print. 

It seems like the positing of E is not handled well or reset to 0 before a print, causing the filament to unload if it was loaded earlier for whatever reason.

**Comment by delwinbest** [comment]
Perhaps include M82 ; Extruder in absolute mode in the start codes?

**Comment by photodude** [comment]
@delwinbest So initially I had the extruders set to M82 in the start gcode and removed that as it blocks the use of the wipe tower. See commit 2b9087478989b4cdfae38dc8d599e1831e66f5e1

I'll have to dig into this a little. I do know that sometimes the printer won't reset properly from one print to the next and requires a reboot for it to reset and home correctly. (prints will start with the center to the left of the actual bed center. only fixed by a reboot)

possibly the distance is an issue in the retract. try the unreleased version with the direct file in github, I shortened the start gcode purge a little. maybe that will help.

**Comment by delwinbest** [comment]
I feel the issue might be with
`G1 Y0 F1200 E0   ; move Y to min endstop and extrude 0 filament`
If the printer does not use absolute for whatever reason (bug??) if would effectively eject if you had previously extruded. 

I'm testing this which seems to work well:

;Start code PrusaSlicer BIBO 2 printers E1 only (i.e. T0)
G21            ; set units to metric
G90            ; absolute positioning
M82 ; Extruder in absolute mode
G92 E0
M107           ; start with the fan off
M190 S{first_layer_bed_temperature[0] - 5} ; wait for bed temp
M140 S[first_layer_bed_temperature]  ; continue bed heating to full temp while other things are happening
M104 S[first_layer_temperature_0] T0 ; set 1st nozzle heater to first layer temperature
M104 S{first_layer_temperature[0] * 0.791} T1 ; set 2nd nozzle heater to 79.1 percent standby temp
M105 ; Report Temperatures
M109 S[first_layer_temperature_0] T0 ; wait for 1st nozzle heat to first layer temperature
M109 S{first_layer_temperature[0] * 0.791} T1 ; wait for 2nd nozzle heat to 79.1 percent standby temp
M105 ; Report Temperatures
G28 X0 Y0        ; move X/Y to min endstops
G28 Z0           ; move Z to min endstops
G28 X0
G1 F1200 E20   ; move Y to min endstop and extrude 0 filament
T[initial_tool]  ; switch to initial tool position
G92 E0.0         ; reset extruder
M117 Cleaning... ; Put Cleaning message on screen, Attempt Nozzle Wipe (for ooze free startup)
G1 X-15.0 Y-92.9 Z0.3 F2400.0     ; move to start-line position
G1 X15.0 Y-92.9 Z0.3 F1000.0 E2   ; draw 1st line
G1 X15.0 Y-92.6 Z0.3 F3000.0      ; move to side a little
G1 X-15.0 Y-92.6 Z0.3 F1000.0 E4  ; draw 2nd line
G1 X-15.0 Y-92.3 Z0.3 F3000.0     ; move to side a little
G1 X15.0 Y-92.3 Z0.3 F1000.0 E6   ; draw 3rd line
G1 X15.0 Y-92 Z0.3 F3000.0        ; move to side a little
G1 X-15.0 Y-92 Z0.3 F1000.0 E8    ; draw 4th line
G1 X-15.0 Y-91.7 Z0.3 F3000.0        ; move to side a little
G1 X15.0 Y-91.7 Z0.3 F1000.0 E10    ; draw 5th line
G1 E12.5 F2400               ; retract filament 0.5mm
G92 E0.0 ; reset extruder and zero the current extruder coordinate before printing
M117 BIBO E1 now Printing... ; Put now printing message on screen

Setting absolute and resetting E seems to work:
`M82 ; Extruder in absolute mode
G92 E0`

I also re-added the 5th cleaning line as well as 20mm extrusion just before the cleaning so that it catches on the build plate and rips away any ooze. Work in progress :)

**Comment by photodude** [comment]
`E20` that doesn't seem right... I need to check something to be sure but I think that should be `E0`

**Comment by photodude** [comment]
@delwinbest Where is the line in the start gcode that says the following
```ini
G1 Y0 F1200 E20   ; move Y to min endstop and extrude 0 filament
```
I can't find that?

Looking at my references and the code I have
```ini
G1 Y0 F1200 E0   ; move Y to min endstop and extrude 0 filament
```

In your code above change 
```
G1 F1200 E20 ; move Y to min endstop and extrude 0 filament
```
from `E20` to `E0` and it should resolve your issue. 

It should be the same as the line in the source start gcode below
```
G1 Y0 F1200 E0 ; move Y to min endstop and extrude 0 filament
```

**Comment by delwinbest** [comment]
Hi Walt, apologies, that was a typo. Please re read the post which I edited.

**Comment by photodude** [comment]
@delwinbest Ok then I'm going to jump to a line I have been suspect too for a while. 
```ini
G1 E12.5 F2400 ; retract filament 0.5mm
```
I think that should read
```ini
G1 E0.5 F2400 ; retract filament 0.5mm
```
so an `E12.5` is definitely not 0.5mm of retraction it's 12.5mm which is a lot (although I'm not totally sure if that should be positive or negative)

I'm still looking into that line.

**Comment by photodude** [comment]
@delwinbest Try completely removing the line following the purge.
```ini
G1 E11.5 F2400 ; retract filament 0.5mm
```

after some testing, I think the purge works best without that line (it seems incorrectly set up too)

I've also updated a number of things in the last two days, a big one is new retraction settings. I also removed some settings which don't apply to BIBO2 machines.

**Comment by InitialJC** [comment]
I am not exactly sure what is going on with this starting Gcode or why it needs to be as complicated as it is however I have gotten much better results by priming the extruder with the following Gcode.

The current release GCode seems to crash my extruder to the left hand side without properly identifying the home/ origin point. The below Gcode will set appropriate absolute position for every print.

G21        ;metric values
G90        ;absolute positioning
M107       ;start with the fan off
G28 X0 Y0  ;move X/Y to min endstops
G28 Z0     ;move Z to min endstops
G1 Z2.0 F400 ;move the platform down 15mm
T0
G92 E0 ; reset extruder
M117 Purge Extruder 1
G1 X-50.0 Y-92.3 Z0.0 F2400.0
G1 X50.0 Y-92.3 Z0.0 F1000.0 E10 ; draw 1st line
G1 X50.0 Y-92 Z0.0 F3000.0 ; move to side a little
G1 X-50.0 Y-92 Z0.0 F1000.0 E20 ; draw 2nd line
G1 E19.5 F3000 ; retract filament 0.5mm
G92 E0 
M117 BIBO Printing...

**Comment by photodude** [comment]
@InitialJC The start gcode is mostly adopted from the BIBO provided Cura start gcode and was originally modified that way based on issues reported in the Facebook BIBO group. One of the big issues with the code you listed above is the lack of temperature preheating of the nozzles and bed which caused failures to print for some people.

There are some things that can be done to simplify the start gcode, but that's probably something to put in a different issue for consideration as there are a lot of things to consider such as how it's implemented for dual extrusion, E1 only, E2 only, Ditto printing printers, etc.

As for your specific left-hand crash issue ... I have noticed that when I'm consecutively printing if I don't move the print head slightly away from the homing switches, before printing occasionally the printer will start off-center.  Doesn't have to be much just enough to keep the switches from being clicked before starting. the off-centered start seems to happen occasionally with cura as well.

**Comment by InitialJC** [comment]
Yes but I'm using the 32 bit w/ UBL and sensor less homing.. the way your starting gcode is currently setup its taking origin resetting it as the starting point and crashing left. Its an order of operations issue. Also for different extruder configurations you can very easily add an addition line from the alternate extruder which is what I do for soluble filaments when printing intricate items.  I haven't ever seen an issue with repetitive printing however with your current starting Gcode its fails and crashes left on every print.. not just a second print.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by photodude** [comment]
@InitialJC Please refer to Issue #3 for your specific 32-bit sensorless homing issue. [https://github.com/photodude/BIBO2-PrusaSlicer/issues/3](https://github.com/photodude/BIBO2-PrusaSlicer/issues/1)

**Comment by photodude** [comment]
@InitialJC It's been confirmed that marlin 2.0.x has a ton of problems with sensorless homing. It's not specific to the start code in the configuration bundle but how marlin is handling sensorless homing.
https://github.com/MarlinFirmware/Marlin/issues?q=is%3Aissue+is%3Aopen+sensorless+homing

**Comment by InitialJC** [comment]
If you want to believe its the firmware thats fine .. but not using your gcode.. and I havent had a single problem with over 40 prints in a row.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by photodude** [comment]
@InitialJC I'm glad you have a solution that's working for you.
Unfortunately, I have no way to replicate or test your issue as I have a fairly stock machine with the original 8-bit motherboard. At the moment all I can do is point to Marlin 2.0.x having a lot of reported issues with sensorless homing. You may want to consider filing a bug report with them and include the start gcode which causes your issue vs the shortened version that you are not having issue with. Maybe something will tip them off on what in the firmware and the start gcode are conflicting and causing the issue.

All the best,

**Comment by photodude** [comment]
@delwinbest I have confirmed that another person with a Stock 8-bit system is having this filament unloading issue. They are using Lou's S3D profile and the start gcode in that. 

Right now we are trying to figure out if the issue is related to the extruder stepper wires, marlin firmware, and/or the start gcode and what specifically is the issue trigger is as I cannot replicate the problem.

**Comment by photodude** [comment]
@delwinbest try putting G92 E0 at the start of the start gcode and removing M82
I'm wondering if a caching issue in the controller start is happening in marlin if the G92 E0 command is not early enough in the start gcode. 

This forum discussion is what I'm basing my suggestion on. 
https://forum.raise3d.com/viewtopic.php?t=1795

The other individual who is also having this issue has M82 in the start gcode, but still has the unloading issue. Their unloading happens between the homing commands at G28 and the move to purge at G1 x-50 Y-92.3 Z0.3 F2400 before the first line of the purge is drawn.

**Comment by photodude** [comment]
@delwinbest Have you tried any of the suggestions or investigated further where in the start gcode the specific error occurs? I am still unable to replicate this issue.

I have another lead on which line might be at fault but as I'm unable to replicate the issue I can't move forward with testing possible errors and fixes.

**Comment by photodude** [comment]
@delwinbest I think I have been able to stumble into some steps which can replicate this "unloading" issue.
1) turn on printer
2) go to "extrusion" from the main menu
3) use the "in" button to advance the filament a significant amount (like set the amount to 10 mm of extrusion and advance 10 times)
4) start a print
5) notice the purge line and skirt do not print as the filament has rewound to the position it started at when the printer was first turned on.
6) stop the print

These steps seem to be consistent with the described behavior from someone else in the community who said the issue occurred between homing and the purge line.

I believe the issue technically comes from `G90 ; absolute positioning` along with  `G28` "home all axis"` which causes the extruders and the XYZ to return to their 0 positions. Since `G92 E0.0 ; reset extruder` comes after `G28` in some profiles (some profiles I used `G28 X0 Y0 ; move X/Y to min endstops` and `G28 Z0 ; move Z to min endstops` before `G92 E0.0 ; reset extruder` which should be less likely to exhibit this behavior than just using `G28`)

Based on my limited testing adding `G92 E0` at the start of the start gcode and before `G90 ; absolute positioning` should be the best method for error proofing the start gcode to prevent the reported issue. This should reset the current extruder position at the start of the print to 0 so when the "home command" `G28` is called the filament won't move since it's already at 0 "home".

I need someone else to test these steps to validate if they can replicate this issue, and repeat the test with the "fix" to ensure the issue has been resolved. Once I have some confirmation of the testing steps and fix validations I'll get a fix entered here and in the Prusaslicer repository.

**Comment by photodude** [comment]
A `G92 E0` must come before  `G1 Y0 F1200 E0`  this will ensure and indeterminate previous extruder position doesn't "rewind" the filament. A fix has been pushed. Closing via 1c61298f0f75a15c0cef88630aec0cd0a01ab246

Sorry, it's taken so long to identify and correct this issue. Version 0.0.8-alpha1 will have this fix as will full released Prusaslicer 0.0.8

**Comment by delwinbest** [comment]
No worries Walt - really appreciate you following up and closing it out. Thanks!
